ext/libdatadog_api/feature_flags.c

```diff
@@ -0,0 +1,382 @@
+  rb_define_method(resolution_details_class, "log?", resolution_details_get_do_log, 0);
+  rb_define_method(resolution_details_class, "flag_metadata", resolution_details_get_flag_metadata, 0);
+
+  // Cache symbol IDs for expected types
+  id_boolean = rb_intern_const("boolean");
+  id_string = rb_intern_const("string");
+  id_number = rb_intern_const("number");
+  id_object = rb_intern_const("object");
+  id_any = rb_intern_const("any");
+  id_integer = rb_intern_const("integer");
+  id_float = rb_intern_const("float");
+}
+
+static VALUE configuration_new(VALUE klass, VALUE json_str) {
+  struct ddog_ffe_Result_HandleConfiguration result = ddog_ffe_configuration_new(borrow_str(json_str));
+  if (result.tag == DDOG_FFE_RESULT_HANDLE_CONFIGURATION_ERR_HANDLE_CONFIGURATION) {
+    rb_raise(feature_flags_error_class, "Failed to create configuration from JSON: %"PRIsVALUE, get_error_details_and_drop(&result.err));
+  }
+  return TypedData_Wrap_Struct(klass, &configuration_data_type, result.ok);
+}
+
+static void configuration_free(void *ptr) {
+  ddog_ffe_Handle_Configuration config = (ddog_ffe_Handle_Configuration)ptr;
+  ddog_ffe_configuration_drop(&config);
+}
+
+static ddog_ffe_ExpectedFlagType expected_type_from_value(VALUE expected_type) {
+  ENFORCE_TYPE(expected_type, T_SYMBOL);
+
+  const ID id = rb_sym2id(expected_type);
+  if (id == id_boolean) {
+    return DDOG_FFE_EXPECTED_FLAG_TYPE_BOOLEAN;
+  } else if (id == id_string) {
+    return DDOG_FFE_EXPECTED_FLAG_TYPE_STRING;
+  } else if (id == id_number) {
+    return DDOG_FFE_EXPECTED_FLAG_TYPE_NUMBER;
+  } else if (id == id_object) {
+    return DDOG_FFE_EXPECTED_FLAG_TYPE_OBJECT;
+  } else if (id == id_any) {
+    return DDOG_FFE_EXPECTED_FLAG_TYPE_ANY;
+  } else if (id == id_integer) {
+    return DDOG_FFE_EXPECTED_FLAG_TYPE_INTEGER;
+  } else if (id == id_float) {
+    return DDOG_FFE_EXPECTED_FLAG_TYPE_FLOAT;
+  } else {
+    rb_raise(feature_flags_error_class, "Internal: Unexpected flag type: %"PRIsVALUE, expected_type);
+  }
+}
+
+// The hash should contain attributes for feature flag evaluation. The
+// special key "targeting_key" (if present) is extracted separately as
+// it has special meaning in the libdatadog API. All other key-value
+// pairs become attributes.
+//
+// Note: all strings are copied into the new EvaluationContext, so
+// there are no safety concerns regarding string lifetimes. However,
+// the caller is responsible for the returned EvaluationContext and
+// must free it when no longer in use.
+static ddog_ffe_Handle_EvaluationContext evaluation_context_from_hash(VALUE hash) {
+  ENFORCE_TYPE(hash, T_HASH);
+
+  const char *targeting_key = NULL;
+  const long max_attr_count = RHASH_SIZE(hash);
+  ddog_ffe_AttributePair *const attrs = ruby_xcalloc(max_attr_count, sizeof(struct ddog_ffe_AttributePair));
+
+  const VALUE keys = rb_funcall(hash, rb_intern_const("keys"), 0);
```

Comment:
This will allocate array, instead we should iterate the hash without an extra allocation